cozy/codegen/rpython.py

```diff
@@ -0,0 +1,422 @@
+"""
+Basic generation of RPython.
+
+The generated code will depend on rlib.
+"""
+
+from contextlib import contextmanager
+from collections import defaultdict
+
+from cozy.codegen.misc import CodeGenerator
+from cozy.syntax_tools import subst
+from cozy.target_syntax import (TBool, TInt,
+        EBool, EEmptyList, ESingleton)
+
+class RPythonPrinter(CodeGenerator):
+
+    def __init__(self, *args, **kwargs):
+        super(RPythonPrinter, self).__init__(*args, **kwargs)
+
+        self.needs = defaultdict(bool)
+        self.instance_names = set()
+
+    @contextmanager
+    def python_block(self, *args):
+        """
+        Write a block delimited in the Python style by a trailing colon.
+        """
+
+        self.begin_statement()
+        self.write(*args, ":")
+        self.end_statement()
+        with self.indented():
+            yield
+
+    @contextmanager
+    def parens(self):
+        self.write("(")
+        yield
+        self.write(")")
+
+    @contextmanager
+    def stmt(self):
+        self.begin_statement()
+        yield
+        self.end_statement()
+
+    def visit_ENum(self, node):
+        self.write(str(node.val))
+
+    def visit_EBool(self, node):
+        self.write("True" if node.val else "False")
+
+    def visit_EEmptyList(self, _):
+        self.write("[]")
+
+    def visit_ESingleton(self, node):
+        self.write("[")
+        self.visit(node.e)
+        self.write("]")
+
+    def visit_EVar(self, node):
+        name = node.id
+        if name in self.instance_names:
+            self.write("self.", name)
+        else:
+            self.write(name)
+
+    def emit_unary(self, op, e):
+        self.needs[op] = True
+        self.write("_", op, "(")
+        self.visit(e)
+        self.write(")")
+
+    def visit_EUnaryOp(self, node):
+        op = node.op
+        if op in ("distinct", "reversed", "sum"):
+            self.emit_unary(op, node.e)
+        elif op == "len":
+            self.write("len(")
+            self.visit(node.e)
+            self.write(")")
+        elif op == "not":
+            self.write("not ")
+            self.visit(node.e)
+        elif op == "empty":
+            self.write("not bool(")
+            self.visit(node.e)
+            self.write(")")
+        elif op == "exists":
+            self.write("bool(")
+            self.visit(node.e)
+            self.write(")")
+        else:
+            import pdb; pdb.set_trace()
+            assert False, "Unhandled unary operation while writing RPython"
```

Comment:
It's a matter of taste, but I would replace these lines with `raise NotImplementedError("unkown operator {!r}".format(op))`.

I generally prefer to keep `pdb` out of committed code. This will be extremely confusing to anyone who accidentally stumbles upon this line.

Reply:
(There are several other references to `pdb` lower down as well.)

Reply:
Ah, yeah, this is a tiny debugging scaffold as I was iterating over this method. It can definitely be removed or replaced.